### PR TITLE
feat(auth): add OAuth2 login

### PR DESCRIPTION
**Pull request by zz5840**
This PR implement the OAuth2 login with GitHub, Google, Microsoft, Discord and OpenID Connect.

See #109.

Currently, this feature is still work in progress, it's not recommended to deploy it to production.

## TODO

- [x] Other platforms (please comment below)
- [x] OIDC
- [x] TOTP
- [x] Error page
- [ ] ~~Revoke~~

## Screenshots

<details>
<summary>Click to view screenshots</summary>

image

image

image

</details>

### CODE REVIEW

**Comment by zz5840** [comment]
Currently GitHub login process in handled manually and Google is handled by `passport-google-oauth20`.

But passport implementation now has security problem since it uses `express-session` to store state param, and it's too heavy to import `express-session` as a dependency, so I'm still hesitating to rewrite it or not.

**Comment by stonith404** [comment]
That's awesome, thanks! Do you think it is possible to create a generic OAuth2 configuration? If this would be possible, the admins could enter the parameters of their own identity provider, e.g., Authentik.

I've been struggling with that too, but it seems like you know OAuth2 better than me. What do you think?

**Comment by zz5840** [comment]
General OAuth2 is impossible, but there is a standard OAuth2 authentication protocol called OpenID Connect (OIDC).

Almost all these platform support OIDC as provider, including Authentik.

- [Authentik](https://goauthentik.io/docs/providers/oauth2/)
- [Authelia](https://www.authelia.com/integration/openid-connect/introduction/)
- [Casdoor](https://casdoor.org/docs/how-to-connect/oidc-client)
- [Keycloak](https://www.keycloak.org/)

So implementing OIDC protocol is enough for these authentication platoforms.

**Comment by stonith404** [comment]
Okay, sorry I've never implemented OAuth2 before so I might make some thinking mistakes. But this is the opportunity to understand it better :)

If I understand it correctly every company implemented OAuth2 a bit differently so it is impossible to make this generic. OIDC is more standardized so this would allow a generic implementation?

E.g if there is a new OIDC provider we don't have to update the code right?

**Comment by zz5840** [comment]
@stonith404 Yes, you are right :)

A standard OAuth procession includes the following steps:

1. Our website redirect user to the provider's login page with our client ID (URL of provider's login page is unknown)
2. After logged in provider's website, the provider will redirect user to a callback URL of our website with a code in query.
3. With the code, our website needs request an access token URL to get access token (URL of access token is also unknown)
4. After getting access token, our website now is able to get user info （but both the URL of user info API and format of user info is unknown)

In summary, it's nearly impossible to implement a generic OAuth2 login.

To solve the problems above, OIDC defines a configuration URL ([example provided by Casdoor](https://door.casdoor.com/.well-known/openid-configuration)), with this URL we will know the necessary URLs during the processing of OAuth authentication (login, access token, user info etc.). Also, OIDC give us an id token (JWT) containing the user info, and it' filed name is fixed ([example](https://casdoor.org/docs/how-to-connect/oidc-client/#oidc-userinfo-fields)). Once we implement OIDC protocol, we'll have compatibility with all OIDC providers.

**Comment by stonith404** [comment]
@zz5840 Ohh I think I got it now. Thanks for explaining :) Would it be possible then to implement OIDC instead of OAuth2 in this case? GitHub and Google sign in can also be used with OIDC, right?

Maybe you know [Immich](https://immich.app), they have the following configuration page for "OAuth":
image

They call it "OAuth" but it's basically OIDC, or I'm wrong?

**Comment by zz5840** [comment]
@stonith404 Oh yes, according to the [docs of Immich](https://immich.app/docs/administration/oauth#overview), they only support OIDC provider. Google does support both [OAuth2](https://developers.google.com/identity/oauth2/web/guides/overview) and [OIDC](https://developers.google.com/identity/openid-connect/openid-connect), but GitHub does not support OIDC.

By the way, OAuth2 is an authorization method, which means access token returned by OAuth is able to access many resources in user's account (for example user's file in Google Drive). But OIDC is just an authentication method, we will only get limited user info if using OIDC.

Since google support OIDC, I will rewrite the google login with OIDC and add general OIDC support in next few days.

**Comment by stonith404** [comment]
Awesome. Sorry the GitHub issue was incorrectly formulated, I hope that you don't have to rewrite much :/

I think we could drop the support for signing in with Github as it would need a custom implementation. Or do you think we should still support it? I'm not sure if many admins actually want to use IDPs that are not self-hosted anyway.

**Comment by zz5840** [comment]
@stonith404 The OAuth login is not only for enterprise, but also for public users. If I deploy this project to my server and make it available to public, I hope users can login with their familiar social account rather than register a new one, so it's necessary for us to adapt as many platforms as possible.

As shown below, Gitea (a self-hosted GitHub alternative) provide a bunch of OAuth2 platforms to chosen from, and many of them do not support OIDC.

image

**Comment by stonith404** [comment]
@zz5840 Ah okay I see, that absolutely makes sense. Thanks!

**Comment by zz5840** [comment]
@stonith404 I'm not so familiar with Nest.js, is it possible to get JWT token with without `@UseGuards(JwtGuard)` (that means JWT token is optional, and I'll handle it manually).

**Comment by zz5840** [comment]
@stonith404 Sorry, I mean I need user (exactly user id)

**Comment by stonith404** [comment]
@zz5840 That's not possible as the user context gets set by the JWT claim `sub`. Without the JWT there is no way the backend knows who requests the route. 

Where exactly would you need this functionality?

**Comment by zz5840** [comment]
> Where exactly would you need this functionality?

Linking account.

I need to judge if user has logged in in callback route. If user has logged in, link their OAuth account to our account, and create a new account if not.

**Comment by stonith404** [comment]
> I need to judge if user has logged in in callback route

Sorry could you elaborate? In the `github/callback` route you check with [this line](https://github.com/stonith404/pingvin-share/pull/276/files#diff-07ae4e1aee6a469b00961baf1dd438e6a4f1c41568cd38e61e6c0dccdb0dccd8R61) if the user has an account already by making a db query with the information provided by the token of GitHub. Why do you need the token additionally?

**Comment by stonith404** [comment]
Oh, or do you mean if the user already has a Pingvin Share account and he wants to link his existing Pingvin Share account from the db with e.g GitHub?

**Comment by zz5840** [comment]
> Oh, or do you mean if the user already has a Pingvin Share account and he wants to link his existing Pingvin Share account from the db with e.g GitHub?

Yes, the following line is just used to login.

https://github.com/stonith404/pingvin-share/blob/c85eaade6a08e278941a7165e55e34923623ccde/backend/src/oauth/oauth.service.ts#L61-L63

**Comment by stonith404** [comment]
Okay I see. You have to parse the jwt manually as far as I know.

In the controller you can get the jwt with `req.cookies.access_token`. Then you create a simple function:
```ts
  async getUserIdFromAccessToken(accessToken: string) {
    try {
      const { sub } = this.jwtService.verify(accessToken);
      return sub;
    } catch {
      // not logged in
      return null;
    }
  }
```

**Comment by zz5840** [comment]
Oh, I see. Thx :)

**Comment by stonith404** [comment]
@zz5840  No problem, just let me know if you need anything :) You can text me on Discord too (stonith404) if this is easier for you. 

But I take a break for today, I had an extremely long day.

**Comment by zz5840** [comment]
@stonith404 Ok, have a good rest.

**Comment by zz5840** [comment]
@stonith404 OAuth processing is contains multiple redirects, so it's hard to show TOTP on login page, maybe we need a new TOTP page instead of integrating it with login page. 
Error page for the same reason.

**Comment by stonith404** [comment]
@zz5840 Yeah you're right this would make more sense.

**Comment by stonith404** [comment]
But does it even make sense to have TOTP when the user signs in with an IDP? Because I can't remember where I was able to specify a 2FA method on an app itself when I've used OAuth because TOTP would get requested by the IDP's login page itself. Or, did I misunderstand something?

**Comment by zz5840** [comment]
@stonith404 I add a config to control whether TOTP should be checked when using OAuth, just leave it to the administrator.

File: backend/src/oauth/oauth.service.ts
Comment (zz5840) [inline backend/src/oauth/oauth.service.ts:96]: @stonith404 the random string doesn't seem to be user friendly, anything advice?

Comment (stonith404) [reply backend/src/oauth/oauth.service.ts:96]: I think we could set password to null if someone registers with OAuth? Then the user can't sign in with email and password which makes sense if you have signed up with an OAuth provider. 
For the username we just could take the content before the `@domain.com` of the email address.

What do you think?

**Comment by stonith404** [comment]
@zz5840 It looks extremely promising, I've just tested it successfully with Keycloak and GitHub and it works with no problems! I'll take a closer look ASAP.

**Comment by stonith404** [comment]
@zz5840  I think it would make more sense to set the password to `null` if you sign up with OAuth instead of setting a random password. I thought about something like this:
image

The user info end would need to return an additional property like `hasPassword`. 

What do you think?

**Comment by zz5840** [comment]
@stonith404 OK, I'll modify it later.

By the way, I think we can put both password and totp into security panel. It's a little messy now.

**Comment by zz5840** [comment]
@stonith404 I think I must do it after you making the password nullable. It occurs some error when I simply edit prisma model. Maybe some changes should apply to the auth code.

**Comment by stonith404** [comment]
@zz5840 I've implemented the nullable password. I also updated the frontend behaviour when the user doesn't have a password. 

Could you please run the formatter when you're done with your changes (`npm run format`)?

**Comment by zz5840** [comment]
@stonith404 Sorry, I will run it in next commit.

**Comment by stonith404** [comment]
@zz5840 No problem, it just could result in merge conflicts because my IDE formats automatically.

**Comment by zz5840** [comment]
@stonith404 Random username and password is fixed, I'll deal with TOTP in following commit. It seems to be a big task. 😫

**Comment by stonith404** [comment]
@zz5840  Yeah this could get a bit difficult. Take your time and let me know if you need anything :) I think after that the feature is ready!

**Comment by zz5840** [comment]
@stonith404 Server will return login token only if username and password have been verified, is it necessary to pass the password and verify the user in `signInTotp` again?

https://github.com/stonith404/pingvin-share/blob/8827ab622fb8647dce6a065e7929607eaa5fe6f2/backend/src/auth/auth.service.ts#L55-L69

**Comment by stonith404** [comment]
Yeah and I see that's not ideal for TOTP. A potential solution would be to store the userId with the [Login Token](https://github.com/stonith404/pingvin-share/blob/8827ab622fb8647dce6a065e7929607eaa5fe6f2/backend/prisma/schema.prisma#L44), then we wouldn't have to enter the email and password again.

**Comment by zz5840** [comment]
@stonith404 OK, I'll solve it.

**Comment by zz5840** [comment]
@stonith404 TOTP works now! Could you please test it?

**Comment by stonith404** [comment]
@zz5840 Great! I'll look into it ASAP

File: backend/src/auth/authTotp.service.ts
Comment (stonith404) [inline backend/src/auth/authTotp.service.ts:10]: There is a runtime error if you use `import QRCode from "qrcode-svg";` instead of `import * as qrcode from "qrcode-svg";`

I don't exactly know why this doesn't work in NestJS but in other frameworks your solution would work.

File: backend/src/auth/authTotp.service.ts
Comment (stonith404) [inline backend/src/auth/authTotp.service.ts:21]: Please remove commented out code

File: frontend/src/components/auth/SignInForm.tsx
Comment (stonith404) [inline frontend/src/components/auth/SignInForm.tsx:105]: Please remove commented out code

**Comment by stonith404** [comment]
Everything looks fine for me. I'm in love with the new TOTP page, it looks awesome :)

I'll test it a bit more over the weekend but after that it should be ready to merge. I built a Docker image if you want to test it on your server too `stonith404/pingvin-share:development`

File: backend/src/auth/authTotp.service.ts
Comment (zz5840) [reply backend/src/auth/authTotp.service.ts:10]: Oh, so wield. I was warned by my IDE that constructors usually start with an uppercase, so I auto-fix it with my IDE.

**Comment by zz5840** [comment]
@stonith404 I will make it ready to merge after I complete the error page today.

**Comment by zz5840** [comment]
@stonith404 The error page is completed, I think it's ready for review now.

So sorry I forgot this is a PR and rebased it to master, there's lots of unrelated commits now.

**Comment by stonith404** [comment]
@zz5840 Awesome, I'll look into it ASAP.

**Comment by stonith404** [comment]
I've just made a minor change that it will display the original error message of the OAuth provider if available:
Screenshot 2023-10-16 at 15 38 20

Even though this message is only in English, it's still better than a generic error message, in my opinion. Would you agree?

**Comment by zz5840** [comment]
@stonith404 Maybe the backend should only response with a i18n key rather than the explicit message. The error message should be written in local files.

**Comment by zz5840** [comment]
According to the documentation of GitHub OAuth App (https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/authorizing-oauth-apps#error-codes-for-the-device-flow), most error code except `access_denied` and `expired_token` is insignificant. So other messages could just return default message. (I will add `expired_token` later)

**Comment by stonith404** [comment]
@zz5840 Yeah that makes sense, thanks! I'll test it again tomorrow and then I'll merge it :)

**Comment by stonith404** [comment]
@zz5840 Sorry, I'm a bit busy. I'll make the final checks over the weekend and merge it afterwards.

**Comment by zz5840** [comment]
@stonith404 OK, I'll do some checking before then. 😃

**Review by stonith404 — APPROVED** [review]
LGTM Thank you very much! Can I merge?

**Comment by zz5840** [comment]
@stonith404 So does it need to be global, I'll delete it if not.
https://github.com/stonith404/pingvin-share/blob/08ffa5481a81e4a098ac0ba895face94616d134a/backend/src/main.ts#L10

**Comment by stonith404** [comment]
@zz5840 Oh yeah, this doesn't need to be global. Because we only return this error page if something went wrong with OAuth, right? For other backend errors we show toasts.

**Comment by zz5840** [comment]
@stonith404 OK, I think we can merge it.

**Comment by stonith404** [comment]
Oh I almost forgot to create a migration. I'll do it quickly

**Comment by stonith404** [comment]
I have created the migration, and I want to test it on my development server. Strangely, the build of the Docker image fails due to connectivity errors with npm. I will try again tomorrow.

**Comment by zz5840** [comment]
Thx, I didn't even know it's necessary to install Python when installing nanoid. 😖

**Pull request merged**
